### PR TITLE
doc: adding adr about avro attrs bridge

### PR DESCRIPTION
**Pull request by jinder1s**
ADR related to #31

Note: This turned into more of a how-to, which will be completed/updated as part of https://openedx.atlassian.net/browse/ARCHBOM-2038.

### CODE REVIEW

**Review by rgraber — COMMENTED** [review]
Overall, this seems more like usage documentation than decisions. Were there any alternatives to any of the decisions listed?

**Comment by jinder1s** [comment]
@ rebecca Yep, this document is kind of a mixture of decision doc, design doc and info dump. Mostly I just wanted to data dump cause there were alot of tiny decisions made during development of AvroAttrsBridge and I wanted them to be specifically listed.

File: docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+              value_serializer=serialize,
+          )
+
+          data
+          # output; {'dt': '2020-05-04T13:37:00'}
+          json.dumps(data)
+          # output:'{"dt": "2020-05-04T13:37:00"}'
+
+2. Each attrs-decorated class has a attrs_attrs class attribute (`source attrs docs <https://www.attrs.org/en/stable/extending.html#extending>`_). It’s a tuple of attrs.Attribute carrying metadata about each attribute.
+   You can get \`type\` info on everything datum defined in an attrs class.
```
[the remark this answers is not in the document]
Comment (jinder1s) [reply docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst:43]: Fixed to suggestion.

**Comment by rgraber** [comment]
> @ rebecca Yep, this document is kind of a mixture of decision doc, design doc and info dump. Mostly I just wanted to data dump cause there were alot of tiny decisions made during development of AvroAttrsBridge and I wanted them to be specifically listed.

Ugh, I forgot this repo doesn't actually have documentation set up. I think it might still be worth it to separate out the 'how to use this tool' documentation for developers who want to use this in the future, and who may not need to know as much about the background or why we chose the things we chose.

**Comment by jinder1s** [comment]
There is already a howto on avro attrs bridge here: https://github.com/eduNEXT/openedx-events/blob/main/docs/howtos/avro_attrs_bridge.rst

I can move the "How to extend bridge" section to that page?

**Comment by jinder1s** [comment]
@rgraber I moved most of the howto stuff into a how to use doc, let me know what you think of the refactoring.

File: docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+The purpose of this ADR is to document decisions made while developing `AvroAttrsBridge <https://github.com/eduNEXT/openedx-events/blob/main/openedx_events/avro_attrs_bridge.py>`_ class.
+
+
+Decision
+--------
+
+Each AvroAttrsBridge class will support:
+
+1. Avro Schema creation is validated at instatiation.
```
Comment (robrap) [inline docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst:23]: Note: I did not do a full review, but just happened to see this typo.
```suggestion
1. Avro Schema creation is validated at instantiation.
```

Comment (rgraber) [inline docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst:23]: Nit: the listen items don't actually complete the sentence at the head of the list, which makes it a bit hard to read. I think we can just scrap the heading and label this "Decisions"
Also the verb tenses of the list items should be consistent

File: docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+- The industry standard for messaging format used with Kafka is Avro (there are other options, but Avro has been the standard for a while). In decision `open-edx-proposals/oep-0052/001-schema-representation.rst <https://github.com/openedx/open-edx-proposals/blob/7bf9acedae5f4290ac2d0e4374c3078278842801/oeps/architectural-decisions/oep-0052/decisions/001-schema-representation.rst>`_: We decided to create a bridge class to convert from Attrs data class to Avro specification.
+
+The purpose of this ADR is to document decisions made while developing `AvroAttrsBridge <https://github.com/eduNEXT/openedx-events/blob/main/openedx_events/avro_attrs_bridge.py>`_ class.
+
+
+Decision
+--------
+
+Each AvroAttrsBridge class will support:
```
Comment (rgraber) [inline docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst:21]: Each subclass? There's only one AvroAttrsBridge class itself

File: docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+AvroAttrsBridge is generalized to serialize/deserialize  basic attrs decorated class. Any specific Kafka requirements will be implemented in KafkaWrapper class, a subclass of AvroAttrsBridge.
+
+
+Open Questions
+--------------
+
+What should go in config?
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+AvroAttrsBridge allows you to config the following values. It provides some default value for testing, but those should not be used in production.
```
Comment (rgraber) [inline docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst:47]: nit: s/config/configure/

Comment (rgraber) [reply docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst:47]: This section is really good! but it seems like it's not actually an Open Question

File: docs/howtos/avro_attrs_bridge.rst [hunk 5]
```diff
@@ -58,3 +61,122 @@ If you want to go from attrs obj to avro dict and back:
    user_data_new_object = user_data_bridge.dict_to_attrs(user_data_dict)
    assert deserealize_user_data == user_data_new_object
 
+
+How to extend AvroAttrsBridge class
+-----------------------------------
+
+At default, attrs.asdict only supports basics types for conversion to dict (Basically, only things you could json.dump). To allow AvroAttrsBridge to work with non-primitive types, a function will be passed to  value_serializer arg in attrs.asdict. The value_serializer function needs to be able to handle any non-primitive types used in an events attrs class.
```
Comment (rgraber) [inline docs/howtos/avro_attrs_bridge.rst:68]: Some typos in this paragraph:
s/basics/basic/
Uncapitalize "Basically"
"a function will be passed to *the* value_serializer arg." also probably should s/will be passed/must be passed/

File: docs/howtos/avro_attrs_bridge.rst [hunk 6]
```diff
@@ -58,3 +61,122 @@ If you want to go from attrs obj to avro dict and back:
...
+
+To make is easier to developers, an extensions interface has been implemented into AvroAttrsBridge.
+To allow AvroAttrsBridge to work with these classes, you can pass in a dict to the extensions keyword to AvroAttrsBridge. The extensions keyword expects a dict with following format: {<type of non-primitive>: <AvroAttrsBridgeExtention subclass for non-primitive>}
+
+The AvroAttrsBridgeExtention subclass should have the following methods:
```
Comment (rgraber) [inline docs/howtos/avro_attrs_bridge.rst:73]: s/Extention/Extension/

File: docs/howtos/avro_attrs_bridge.rst [hunk 7]
```diff
@@ -58,3 +61,122 @@ If you want to go from attrs obj to avro dict and back:
...
+
+Handling Schema Evolution
+-------------------------
+
+If an attrs decorated class has a default value for one of its attributes, avro_attrs_bridge will assume that attribute is optional. This is to allow attrs events to change over time. If you want to add a new attribute to old attrs decorated class, please set a default value for it so that data created using old version can still be read.
+
+This has not been tested in production. If you do some testing, please update this and create further how_tos to handle schema evolution.
+
+
+context
```
Comment (rgraber) [inline docs/howtos/avro_attrs_bridge.rst:97]: Capitalize

File: docs/howtos/avro_attrs_bridge.rst [hunk 8]
```diff
@@ -58,3 +61,122 @@ If you want to go from attrs obj to avro dict and back:
...
    assert deserealize_user_data == user_data_new_object
 
+
+How to extend AvroAttrsBridge class
+-----------------------------------
+
+At default, attrs.asdict only supports basics types for conversion to dict (Basically, only things you could json.dump). To allow AvroAttrsBridge to work with non-primitive types, a function will be passed to  value_serializer arg in attrs.asdict. The value_serializer function needs to be able to handle any non-primitive types used in an events attrs class.
+
+To make is easier to developers, an extensions interface has been implemented into AvroAttrsBridge.
+To allow AvroAttrsBridge to work with these classes, you can pass in a dict to the extensions keyword to AvroAttrsBridge. The extensions keyword expects a dict with following format: {<type of non-primitive>: <AvroAttrsBridgeExtention subclass for non-primitive>}
```
Comment (rgraber) [inline docs/howtos/avro_attrs_bridge.rst:71]: I don't understand "pass in a dict to the extensions keyword." How does one pass a dict to a keyword? Did you mean attribute?

File: docs/howtos/avro_attrs_bridge.rst [hunk 9]
```diff
@@ -58,3 +61,122 @@ If you want to go from attrs obj to avro dict and back:
...
+
+3. Avro deals with conversion between "dict" like objects to bytes.
+
+   So our solution needs to go from attrs decorated classes to "dict" like objects
+
+3. A schema must be used to deserialize encoded data.
+
+   The encoded data does not include type or field names. To read the data, the schema used to read the data must be identical to the schema used to write data.
+
+4. evolution requirements
```
Comment (rgraber) [inline docs/howtos/avro_attrs_bridge.rst:176]: Capitalize

File: docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+2. Allows for following conversions:
+
+   attrs decorated class object => dict => Avro encoded string => dict => attrs decorated class object
+
+3. Support doing the above by default for all attrs decorated classes in openedx-events repository
+
+4. Provide ability to extend AvroAttrsBridge to support any attrs decorated classes outside of openedx-events repository
+
+5. Follow cloudevents specification as stated in `OEP-41`_: Asynchronous Server Event Message Format.
```
Comment (rgraber) [inline docs/decisions/0004-avro-attrs-bridge-implementation-decisions.rst:35]: capitalize CloudEvents

**Review by rgraber — COMMENTED** [review]
Many nits, some content things

**Comment by robrap** [comment]
Closing for now. This turned into more of a how-to, which should be completed/updated as part of https://openedx.atlassian.net/browse/ARCHBOM-2038.

**Pull request closed without merging**